microprofile/WFLY-13640_MicroProfile_Reactive_Messaging.adoc

```diff
@@ -0,0 +1,183 @@
+
+* https://github.com/smallrye/smallrye-reactive-messaging[Reactive Messaging]
+* https://github.com/smallrye/smallrye-mutiny/tree/master/reactive-streams-operators[RSO]
+
+Only supporting MicroProfile Reactive Messaging 1.0 means there is *no* support for the https://github.com/smallrye/smallrye-reactive-messaging/blob/master/api/src/main/java/org/eclipse/microprofile/reactive/messaging/Channel.java[@Channel] annotation and the https://github.com/smallrye/smallrye-reactive-messaging/blob/master/api/src/main/java/org/eclipse/microprofile/reactive/messaging/Emitter.java[Emitter] interface. These are essential to be allow user/imperative code push data into, and get data from, the reactive messaging channels. So this initial iteration will not be able to bridge user initiated code (such as a REST API call) and the streams used by Reactive Messaging. The processors and subscribers for Reactive Messaging may however 'call out' and for example save received data in a database.
+
+There are several libraries all implementing the http://www.reactive-streams.org[Reactive Streams] specification. One new implementation is https://github.com/smallrye/smallrye-mutiny[SmallRye Mutiny] which is used under the hood by the SmallRye implementations of RSO and Reactive Messaging. We already use https://github.com/ReactiveX/RxJava/tree/2.x[RxJava 2] as an internal (i.e. it is a private module) dependency for the runtime functionality of some existing WildFly extensions. However, these should not be exposed to users at this time. We should settle on the implementation of RSO, which, although has less functionality than the other two mentioned, is a MicroProfile specification.
+
+In addition for more 'finite' sets of data, Java's CompletionStage is supported.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-13640
+* https://issues.redhat.com/browse/EAP7-1405
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1518[Quickstart]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:jstourac@redhat.com[Jan Stourac]
+* mailto:okotek@redhat.com[Ondrej Kotek]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [X] QE
+
+=== Affected Projects or Components
+The changes will be made in WildFly adding a subsystem for each spec.
+
+The functionality of these subsystems will be provided by the SmallRye implementations and their dependencies.
+
+In addition for OpenShift we need to make sure that provides out-of-the box integration with the Kafka managed service (by leveraging integration at the operator level). This  integration is scheduled post-WildFly 23. My current understanding is that we should be able to use environment variables and system properties to indicate to the user application the address and ports of the Kafka managed service.
```

Comment:
@jmesnil Does this look ok as we discussed yesterday?

Reply:
I don't know enough about Kafka on OpenShift to confirm this.

If I read https://strimzi.io/docs/operators/latest/quickstart.html#proc-using-amq-streams-str correctly, EAP will need to know the IP address and port to connect to Kafka with something like: `kafka.bootstrap.servers=localhost:9092`  (from [Quarkus Kafka Guide](https://quarkus.io/guides/kafka))

So when we start the EAP application image, we need to setup an environment variable for it.
The values for the IP address and port can be deduced by looking at the `KafkaTopic` and `Kafka` CRD and check a matching listener.

So if the EAP application is using "my-cluster" topic, we have to look at the `KafkaTopic` with the name "my-topic".
Then in that resource, we look at the `strimzi.io/cluster` label to find the name of the `Kafka` resource that manages this topic.
Finally, we need to find a matching listeners from this `Kafka` resource: 

```
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    replicas: 1
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
        ...
```

For the IP address, we have to find the service that exposes the matching listeners and determine its IP address.
Once we have that, we have all the info we need to create the `kafka.bootstrap.servers` config property and run EAP.

If we want to hide all this info seeking and service binding from the user, I envision the EAP Operator doing everything for you as long as you tell it that you want to communicate with the `my-topic` topic:

```
spec:
  kafka:
    topic: "my-topic"
    # to be able to determine which listener (and port) to connect to)
    listener: "plain"